File: c15_string_sets/L14.cpp
//
// Created by 李恒 on 2020/1/3.
//

// 14 最长公共前缀

#include <iostream>
#include <cstring>
#include <vector>
using namespace std;

class Solution {
public:
    string longestCommonPrefix(vector<string>& strs) {
        if(strs.size() == 0)
            return "";
        string res_str = "";

        int tem_s = strs[0].size();
        for(int i=1 ;i < strs.size();i++){
            if (strs[i].size() > tem_s)
                tem_s = strs[i].size();
        }

        for(int i=0; i<tem_s;i++){
            char tem_str = strs[0][i];
            bool flag = true ;
            for(int j=1 ;j < strs.size();j++){
                if (strs[j][i] != tem_str){
                    flag = false;
                    break;
                }
            }
            if(flag)
                res_str += tem_str;
            else
                break;
        }
        return res_str;
    }
};

File: python-algorithm/ds_6_string/L14.py
# -*- coding: utf-8 -*-


class Solution:
    def longestCommonPrefix(self, strs: List[str]) -> str:
        if strs == []:
            return ""
        res = ""
        min_len = inf
        for word in strs:
            min_len = min(min_len, len(word))
        for i in range(min_len):
            ref = strs[0][i]
            flag = 1
            for j in range(len(strs)):
                if strs[j][i] != ref :
                    flag = 0 # 说明遍历过程中找到了不同的字符
                    break # 注意break只能终止当前层的循环，无法终止外部层的循环
            if flag:
                res += ref
            else:
                break
        return res

# 巧妙借助python内部的字符串排序
class Solution:
    def longestCommonPrefix(self, strs):
        if not strs: return ""
        s1 = min(strs)
        s2 = max(strs)
        for i,x in enumerate(s1):
            if x != s2[i]:
                return s2[:i]
        return s1